src/libserver/task.h

```diff
@@ -51,13 +51,16 @@ enum rspamd_task_stage {
 	RSPAMD_TASK_STAGE_CLASSIFIERS_POST = (1u << 8u),
 	RSPAMD_TASK_STAGE_COMPOSITES = (1u << 9u),
 	RSPAMD_TASK_STAGE_POST_FILTERS = (1u << 10u),
-	RSPAMD_TASK_STAGE_LEARN_PRE = (1u << 11u),
-	RSPAMD_TASK_STAGE_LEARN = (1u << 12u),
-	RSPAMD_TASK_STAGE_LEARN_POST = (1u << 13u),
-	RSPAMD_TASK_STAGE_COMPOSITES_POST = (1u << 14u),
-	RSPAMD_TASK_STAGE_IDEMPOTENT = (1u << 15u),
-	RSPAMD_TASK_STAGE_DONE = (1u << 16u),
-	RSPAMD_TASK_STAGE_REPLIED = (1u << 17u)
+	RSPAMD_TASK_STAGE_AUTOLEARN_PRE = (1u << 11u),
```

Comment:
I'm not convinced that we need 3 extra stages for that. We can probably move the whole logic to `STAGE_LEARN_PRE` and execute it unconditionally.

Reply:
Correct me if I am wrong, but as I see it, we need 2 stages to decide if we can autolearn.

As we only get the flags from `cache_check` asynchronously, since it is a Lua callback that sets them.
There is code in `cache_check` that sets `RSPAMD_TASK_FLAG_UNLEARN` but as far as I see it can only be reached with the sqlite3 backend which has code to return  `RSPAMD_LEARN_UNLEARN`.

So if `STAGE_LEARN_PRE` does the cache check. We only get the symbols in the next stage (unless there is a way to get the Flags of the LUA callback earlier).
So the decision on whether we do autolearn or not would have to move into `STAGE_LEARN` which would also mean it has to be called unconditionally (at least for the first part of the stage).